### PR TITLE
[GCS]Add in-memory gcs table storage

### PR DESCRIPTION
**Pull request by ffbin**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by wumuzi520** [comment]
LGTM!

File: src/ray/gcs/gcs_server/test/gcs_table_storage_test_base.h [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+  void WaitPendingDone() { WaitPendingDone(pending_count_); }
+
+  void WaitPendingDone(std::atomic<int> &pending_count) {
+    auto condition = [&pending_count]() { return pending_count == 0; };
+    EXPECT_TRUE(WaitForCondition(condition, wait_pending_timeout_.count()));
+  }
+
+ protected:
+  std::shared_ptr<rpc::JobTableData> GenJobTableData(JobID job_id) {
```
Comment (wumuzi520) [inline src/ray/gcs/gcs_server/test/gcs_table_storage_test_base.h:125]: Put it to gcs_test_util.h?

File: src/ray/gcs/gcs_server/test/gcs_table_storage_test_base.h [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    auto job_table_data = std::make_shared<rpc::JobTableData>();
+    job_table_data->set_job_id(job_id.Binary());
+    job_table_data->set_is_dead(false);
+    job_table_data->set_timestamp(std::time(nullptr));
+    job_table_data->set_driver_ip_address("127.0.0.1");
+    job_table_data->set_driver_pid(5667L);
+    return job_table_data;
+  }
+
+  std::shared_ptr<rpc::ActorTableData> GenActorTableData(const JobID &job_id) {
```
Comment (wumuzi520) [inline src/ray/gcs/gcs_server/test/gcs_table_storage_test_base.h:135]: ditto.

File: src/ray/gcs/gcs_server/test/redis_gcs_table_storage_test.cc [hunk 3]
```diff
@@ -15,111 +15,34 @@
...
+  void InitTableStorage() override {
     gcs::RedisClientOptions options("127.0.0.1", REDIS_SERVER_PORT, "", true);
     redis_client_ = std::make_shared<gcs::RedisClient>(options);
     RAY_CHECK_OK(redis_client_->Connect(io_service_pool_->GetAll()));
 
     gcs_table_storage_ = std::make_shared<gcs::RedisGcsTableStorage>(redis_client_);
   }
 
-  void DisconnectStoreClient() override { redis_client_->Disconnect(); }
+  void DeInitTableStorage() override { redis_client_->Disconnect(); }
```
Comment (wumuzi520) [inline src/ray/gcs/gcs_server/test/redis_gcs_table_storage_test.cc:37]: Put it to TearDown?

File: src/ray/gcs/gcs_server/test/redis_gcs_table_storage_test.cc [hunk 4]
```diff
@@ -15,111 +15,34 @@
...
-  GcsTableStorageTest() {}
-
-  virtual ~GcsTableStorageTest() {}
-
   static void SetUpTestCase() { RedisServiceManagerForTest::SetUpTestCase(); }
 
   static void TearDownTestCase() { RedisServiceManagerForTest::TearDownTestCase(); }
 
-  void InitStoreClient() override {
+  void InitTableStorage() override {
```
Comment (wumuzi520) [inline src/ray/gcs/gcs_server/test/redis_gcs_table_storage_test.cc:29]: Put it to SetUp?

Comment (ffbin) [reply src/ray/gcs/gcs_server/test/redis_gcs_table_storage_test.cc]: fixed

File: src/ray/gcs/gcs_server/test/redis_gcs_table_storage_test.cc (hunk 3 above)
Comment (ffbin) [reply src/ray/gcs/gcs_server/test/redis_gcs_table_storage_test.cc]: fixed

File: src/ray/gcs/gcs_server/test/gcs_table_storage_test_base.h (hunk 2 above)
Comment (ffbin) [reply src/ray/gcs/gcs_server/test/gcs_table_storage_test_base.h]: fixed

File: src/ray/gcs/gcs_server/test/gcs_table_storage_test_base.h (hunk 1 above)
Comment (ffbin) [reply src/ray/gcs/gcs_server/test/gcs_table_storage_test_base.h]: fixed

File: BUILD.bazel [hunk 5]
```diff
@@ -807,6 +809,17 @@ cc_test(
     ],
 )
 
+cc_library(
+    name = "gcs_test_util_lib",
```
Comment (wumuzi520) [inline BUILD.bazel:813]: I think it is unnecessary to define this lib as it just include a simple `gcs_test_util.h`.

File: BUILD.bazel [hunk 6]
```diff
@@ -807,6 +809,17 @@ cc_test(
     ],
```
Comment (wumuzi520) [inline BUILD.bazel:809]: add ":gcs_test_util_lib",

Comment (ffbin) [reply BUILD.bazel:824]: fixed

**Review by wumuzi520 — APPROVED** [review]

**Review by raulchen — APPROVED** [review]

**Pull request merged**